File: aexample/src/main/scala/org/judal/examples/scala/jdbc/E06_CreateTablesFromDDLXML.scala
package org.judal.examples.scala.jdbc

import org.junit.Test
import org.scalatest.Suite

import org.judal.storage.DataSource

import org.judal.jdbc.metadata.SQLXmlMetadata
import org.judal.jdbc.JDBCRelationalDataSource

import org.judal.examples.Resources.getResourceAsStream

import collection.JavaConverters._

import scala.collection.JavaConversions._

/**
 * Example of how to create tables read from a DDL XML file
 * @see https://db.apache.org/ddlutils/schema/
 */
class E06_CreateTablesFromDDLXML extends Suite {

	@Test def demo() = {

		val dataSource = E02_CreateAdditionalDataSource.create()
				
		E06_CreateTablesFromDDLXML.createSchemaObjects(dataSource)
		
		E06_CreateTablesFromDDLXML.dropSchemaObjects(dataSource)
		
		E02_CreateAdditionalDataSource.close(dataSource)
	}
		
}

object E06_CreateTablesFromDDLXML {

	def createSchemaObjects(dataSource: JDBCRelationalDataSource) = {		
		val xmlMeta = new SQLXmlMetadata (dataSource)

		// Load table definitions from a DDL XML file
		val metadata = xmlMeta.readMetadata(getResourceAsStream("ddlmetadata.xml"))

		val options = Map[String,Object](
		    DataSource.CATALOG -> "PUBLIC",
		    DataSource.SCHEMA -> "PUBLIC" ).asJava
		
		metadata.tables.foreach { tdef => dataSource.createTable(tdef, options) }
		    			
		metadata.indexes.foreach { idef => dataSource.createIndex(idef) }
	}

	def dropSchemaObjects(dataSource: JDBCRelationalDataSource) = {
		
		val metadata = new SQLXmlMetadata (dataSource).readMetadata(getResourceAsStream("ddlmetadata.xml"))

		metadata.tables.toList.reverse.foreach { tdef => dataSource.dropTable(tdef.getName, false) }
		
	}
  
}

File: aexample/src/main/java/org/judal/examples/java/jdbc/E06_CreateTablesFromDDLXML.java
package org.judal.examples.java.jdbc;

import org.junit.Test;
import org.judal.metadata.IndexDef;
import org.judal.metadata.SchemaMetaData;
import org.judal.metadata.TableDef;
import org.judal.storage.DataSource;
import org.judal.jdbc.metadata.SQLXmlMetadata;
import org.judal.jdbc.JDBCRelationalDataSource;

import static org.judal.examples.Resources.getResourceAsStream;

import java.io.IOException;
import java.util.HashMap;
import java.util.LinkedList;
import java.util.Map;

import javax.jdo.JDOException;

/**
 * Example of how to create tables read from a DDL XML file
 * @see https://db.apache.org/ddlutils/schema/
 */
public class E06_CreateTablesFromDDLXML {

	@Test
	public void demo() throws Exception {

		JDBCRelationalDataSource dataSource = E02_CreateAdditionalDataSource.create();
				
		createSchemaObjects(dataSource);
		
		dropSchemaObjects(dataSource);
		
		E02_CreateAdditionalDataSource.close(dataSource);
	}
	
	public static void createSchemaObjects(JDBCRelationalDataSource dataSource) throws JDOException, IOException {
		
		SQLXmlMetadata xmlMeta = new SQLXmlMetadata (dataSource);

		// Load table definitions from a DDL XML file
		SchemaMetaData metadata = xmlMeta.readMetadata(getResourceAsStream("ddlmetadata.xml"));

		Map<String,Object> options = new HashMap<>();
		options.put(DataSource.CATALOG, dataSource.getProperties().getOrDefault(DataSource.CATALOG, ""));
		options.put(DataSource.SCHEMA, dataSource.getProperties().getOrDefault(DataSource.CATALOG, "PUBLIC"));
		
		// Create tables in the database
		for (TableDef tdef : metadata.tables())
			dataSource.createTable(tdef, options);

		// Create indexes in the database
		for (IndexDef idef : metadata.indexes())
			dataSource.createIndex(idef);		
	}

	public static void dropSchemaObjects(JDBCRelationalDataSource dataSource) throws JDOException, IOException {
		
		SQLXmlMetadata xmlMeta = new SQLXmlMetadata (dataSource);
		
		SchemaMetaData metadata = xmlMeta.readMetadata(getResourceAsStream("ddlmetadata.xml"));
		
		// Reverse the table list
		LinkedList<String> tableNames = new LinkedList<>();
		for (TableDef tdef : metadata.tables())
			tableNames.addFirst(tdef.getName());
		
		// Drop tables in the database
		for (String tableName: tableNames)
			dataSource.dropTable(tableName, false);
	}
	
}
